### ISSUE TITLE
Running Cloudformation module asynchronously does not work

### ISSUE THREAD

**Issue by tveon** [ansible/ansible-modules-core#3122]
##### Issue Type:
- Bug Report
##### Plugin Name:
- cloudformation
##### Ansible Version:

```
ansible 2.0.0.2
  config file = /Users/tvtr/.ansible.cfg
  configured module search path = Default w/o overrides
```
##### Ansible Configuration:

```
[defaults]
ANSIBLE_COW_SELECTION=small
host_key_checking=false
```
##### Environment:

Running locally on OSX El Capitan (10.11.3)
##### Summary:

Since Cloudformation often takes tens of minutes to create the desired resources, it is annoying to have to wait for, but when trying to execute a cloudformation-task asynchronously I experience two problems:
1. The stack is not created
2. When checking back on the task, I get the error:

```
fatal: [localhost]: FAILED! => {"failed": true, "msg": "ERROR! The conditional check 'job_result.finished' failed. The error was: ERROR! error while evaluating conditional (job_result.finished): ERROR! 'dict object' has no attribute 'finished'"}
```
##### Steps To Reproduce:

(AWS credentials needs to be configured.)

Run the following playbook in the same folder as the following cloudformation-template

```

---
- name: cloudformation-bug
  hosts: localhost
  connection: local

  tasks:

  - name: Cloudformation task
    cloudformation:
      stack_name:   Test
      state:        absent
      region:       eu-west-1
      template:     cfn-test-template.json
    async: 100
    poll: 0
    register: sleeper

  - name: Another long running task
    command: sleep 3
    async: 100
    poll: 0
    register: sleeper2

  - name: Check back on last first
    async_status: jid="{{ sleeper2.ansible_job_id }}"
    register: job_result
    until: job_result.finished
    retries: 10

  - name: Check back on cloudformation
    async_status: jid="{{ sleeper.ansible_job_id }}"
    register: job_result
    until: job_result.finished
    retries: 10
```

Cloudformation-template

```
{
  "AWSTemplateFormatVersion" : "2010-09-09",

  "Resources" : {
    "Noop" : {
      "Type" : "AWS::CloudFormation::WaitConditionHandle"
    }
  }
}
```
##### Expected Results:
1. A cloudformation-stack with the name "Test" will be created
2. It would be possible to get the status later
##### Actual Results:
1. No stack is created
2. The check-in fails - as described above

```
tvtr::imac-dgknv0cnf8jc { ~/IdeaProjects/devtest }$ ansible-playbook bugreport.yml -vvvv
Using /Users/tvtr/.ansible.cfg as config file
 [WARNING]: provided hosts list is empty, only localhost is available

Loaded callback default of type stdout, v2.0
1 plays in bugreport.yml

PLAY [cloudformation-bug] ******************************************************

TASK [setup] *******************************************************************
ESTABLISH LOCAL CONNECTION FOR USER: tvtr
127.0.0.1 EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420726.37-5789713416591 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420726.37-5789713416591 )" )
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmpgCaZYR TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.37-5789713416591/setup
127.0.0.1 EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.37-5789713416591/setup; rm -rf "/Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.37-5789713416591/" > /dev/null 2>&1
ok: [localhost]

TASK [Cloudformation task] *****************************************************
task path: /Users/tvtr/IdeaProjects/devtest/bugreport.yml:8
ESTABLISH LOCAL CONNECTION FOR USER: tvtr
127.0.0.1 EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923 )" )
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmp0EOz7r TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923/cloudformation
127.0.0.1 EXEC chmod a+rx /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923/cloudformation
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmpezSR_o TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923/async_wrapper
127.0.0.1 EXEC chmod a+rx /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923/async_wrapper
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmpRXIQb_ TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923/arguments
127.0.0.1 EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923/async_wrapper 920045905259 100 /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923/cloudformation /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923/arguments
127.0.0.1 EXEC rm -f -r /Users/tvtr/.ansible/tmp/ansible-tmp-1456420726.64-176620018956923/ > /dev/null 2>&1
ok: [localhost] => {"ansible_job_id": "920045905259.8120", "changed": false, "results_file": "/Users/tvtr/.ansible_async/920045905259.8120", "started": 1}

TASK [Another long running task] ***********************************************
task path: /Users/tvtr/IdeaProjects/devtest/bugreport.yml:18
ESTABLISH LOCAL CONNECTION FOR USER: tvtr
127.0.0.1 EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809 )" )
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmp3fshai TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809/command
127.0.0.1 EXEC chmod a+rx /Users/tvtr/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809/command
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmpfap4gJ TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809/async_wrapper
127.0.0.1 EXEC chmod a+rx /Users/tvtr/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809/async_wrapper
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmp9dbPeu TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809/arguments
127.0.0.1 EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /Users/tvtr/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809/async_wrapper 728828658006 100 /Users/tvtr/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809/command /Users/tvtr/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809/arguments
127.0.0.1 EXEC rm -f -r /Users/tvtr/.ansible/tmp/ansible-tmp-1456420727.76-96601870142809/ > /dev/null 2>&1
ok: [localhost] => {"ansible_job_id": "728828658006.8138", "changed": false, "results_file": "/Users/tvtr/.ansible_async/728828658006.8138", "started": 1}

TASK [Check back on last first] ************************************************
task path: /Users/tvtr/IdeaProjects/devtest/bugreport.yml:24
ESTABLISH LOCAL CONNECTION FOR USER: tvtr
127.0.0.1 EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420728.9-12992414257137 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420728.9-12992414257137 )" )
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmpl_GpsQ TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420728.9-12992414257137/async_status
127.0.0.1 EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/tvtr/.ansible/tmp/ansible-tmp-1456420728.9-12992414257137/async_status; rm -rf "/Users/tvtr/.ansible/tmp/ansible-tmp-1456420728.9-12992414257137/" > /dev/null 2>&1
FAILED - RETRYING: TASK: Check back on last first (9 retries left). Result was: {u'ansible_job_id': u'728828658006.8138', u'started': 1, u'changed': False, u'finished': 0, u'results_file': u'/Users/tvtr/.ansible_async/728828658006.8138', 'invocation': {'module_name': u'async_status', u'module_args': {u'jid': u'728828658006.8138', u'mode': u'status'}}}
127.0.0.1 EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420733.99-117041445087675 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420733.99-117041445087675 )" )
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmpm44PIi TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420733.99-117041445087675/async_status
127.0.0.1 EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/tvtr/.ansible/tmp/ansible-tmp-1456420733.99-117041445087675/async_status; rm -rf "/Users/tvtr/.ansible/tmp/ansible-tmp-1456420733.99-117041445087675/" > /dev/null 2>&1
changed: [localhost] => {"ansible_job_id": "728828658006.8138", "changed": true, "cmd": ["sleep", "3"], "delta": "0:00:03.009859", "end": "2016-02-25 18:18:50.912531", "finished": 1, "invocation": {"module_args": {"_raw_params": "sleep 3", "_uses_shell": false, "chdir": null, "creates": null, "executable": null, "removes": null, "warn": true}, "module_name": "async_status"}, "rc": 0, "start": "2016-02-25 18:18:47.902672", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}

TASK [Check back on cloudformation] ********************************************
task path: /Users/tvtr/IdeaProjects/devtest/bugreport.yml:30
ESTABLISH LOCAL CONNECTION FOR USER: tvtr
127.0.0.1 EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420734.12-279891892298093 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1456420734.12-279891892298093 )" )
127.0.0.1 PUT /var/folders/l5/bxg1l6tj03gdm2s4y8kv5_pr9m08jd/T/tmpp3xQ5S TO /Users/tvtr/.ansible/tmp/ansible-tmp-1456420734.12-279891892298093/async_status
127.0.0.1 EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/tvtr/.ansible/tmp/ansible-tmp-1456420734.12-279891892298093/async_status; rm -rf "/Users/tvtr/.ansible/tmp/ansible-tmp-1456420734.12-279891892298093/" > /dev/null 2>&1
fatal: [localhost]: FAILED! => {"failed": true, "msg": "ERROR! The conditional check 'job_result.finished' failed. The error was: ERROR! error while evaluating conditional (job_result.finished): ERROR! 'dict object' has no attribute 'finished'"}

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=1    unreachable=0    failed=1   
```

**Comment by ahawkins** [comment]
I'm seeing the same error with a playbook like so:

```
    - name: "Pull images"
      command: "sandbox pull"
      with_items: "{{ sandboxes }}"
      environment: "{{ sandbox_env|combine({'SANDBOX_NAME':'{{ item }}'}) }}"
      register: sandbox_pulls
      async: 900 # 15 Minutes
      poll: 0 # Fire and forget
      become: yes
      become_user: sandbox

    - name: "Wait for pulls to complete"
      async_status: jid={{ item.ansible_job_id }}
      register: pull_jobs
      until: pull_jobs.finished
      retries: 300
      with_items: "{{ sandbox_pulls.results }}"
```

**Comment by do3cc** [comment]
@ahawkins I had the same problem. Since lots of -v did not give me the dict I needed, I grepped for the exception error message and added the the all_vars attribute to the error.
This way I found out, that he didn't know the ansible_job_id, so an error got registered.
In my case, the first task was delegated_to another host, while the secon'd wasn't.
Maybe in your case it works if you add become_user and become.

**Comment by mdaize** [comment]
I have a number of CloudFormation scripts that can each take 10-15minutes to run. This means that a single ansible task can take several HOURS depending on how many are run. Async is a must in this situation.

**Comment by ansibot** [comment]
@jsmartin, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by jsmartin** [comment]
Finally hit this bug today.  No idea what is the root cause.  I don't think you have to do special coding to get async to work with modules.  Perhaps the core maintainers can chime in here.

**Comment by ansibot** [comment]
@jsmartin, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jsmartin** [comment]
Digging deeper into this issue... I register the results of an async'd cloudformation task.

I noticed as part of the output is the "results_file": "/Users/jmartin/.ansible_async/15185602322.22898",

So I opened that file and see:

{"msg": "Traceback (most recent call last):\n  File \"/Users/jmartin/.ansible/tmp/ansible-tmp-1472246721.12-25785912081488/async_wrapper\", line 90, in _run_module\n    result = json.loads(outdata)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py\", line 338, in loads\n    return _default_decoder.decode(s)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py\", line 366, in decode\n    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py\", line 384, in raw_decode\n    raise ValueError(\"No JSON object could be decoded\")\nValueError: No JSON object could be decoded\n", "failed": 1, "cmd": "/Users/jmartin/.ansible/tmp/ansible-tmp-1472246721.12-25785912081488/cloudformation", "data": "", "ansible_job_id": "15185602322.22898"}

So something is not returning valid JSON.

**Comment by nitzmahone** [comment]
You might be running into similar issues as https://github.com/ansible/ansible/issues/17035.

There's an output scrubber in the normal module exec action that can filter garbage from stdout to find just the JSON module return, but async doesn't currently have access to it, so any badly-behaved module that has extra "junk" in stdout will fail this way under async (since async reads the JSON from stdout directly and adds stuff like "finished"). It's a little suspect that the stack isn't getting created though, as that implies the module failed in some other way.

**Comment by jsmartin** [comment]
@nitzmahone the stacks get created just fine when they are not run in async mode.

I did some further digging and ran  the "/Users/jmartin/.ansible/tmp/ansible-tmp-1472246721.12-25785912081488/cloudformation" file that the async_wrapper contained, and it created the stack.  So I don't think there's a cloudformation module problem.  Something higher up the chain.

**Comment by nitzmahone** [comment]
Yeah, I just meant that it might _not_ be the most typical async issue of "module ran fine but extra junk in stdout prevented retval parsing". IIRC I've heard complaints of async issues with things like `ansible_python_interpreter` too, so if you're running from a venv or something else where you're not always using the system python version, that might be an avenue to explore. Also curious if it works better on 2.1.x or devel (IIRC there were some recent fixes to async in devel).

**Comment by jsmartin** [comment]
Tried in both devel and 2.1.1, neither work.  I am running it inside a venv.  I've tried it with and without ansible_python_interpeter too.  No affect.

**Comment by jsmartin** [comment]
Seems to be more like these issues:  https://github.com/ansible/ansible/issues/13421 and https://github.com/ansible/ansible/issues/16724

**Comment by jsmartin** [comment]
This is definitely a case of ansible not using the specified interpreter.  Below is a snippet from the file `/Users/jmartin/.ansible_async/15185602322.22898`. As you can see, the first message is that boto is not found.  I can assure you that it's in my running virtual env.  I even went so far as hard coding the absolute path the python interpreter to the hosts file to no avail.

`{"msg": "boto required for this module", "failed": true`, ..snip }

**Comment by jsmartin** [comment]
It appears that when the cloudformation module is invoked in async mode and the template is passed with a relative path, the template cannot be found.  

I discovered this by digging through the async_wrapper module and adding some debug commands after line 81:

```
    outdata = ''
    logger = open('/tmp/alog', 'a')
    logger.write("\nBEGIN" + "\n")
    logger.write(tmp_job_path + "\n")
    cmd = shlex.split(wrapped_cmd )
    logger.write("WRAPPED " + wrapped_cmd + "\n")
    logger.write("UNWRAPPED " + cmd + "\n")
    script = subprocess.Popen(wrapped_cmd, shell=False, stdin=None, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    (outdata, stderr) = script.communicate()
    logger.write("RAN SCRIPT\n")
    logger.write("STDERR " + stderr + "\n")
    logger.write("STDOUT " + stdout + '"n")
    logger.close()
```

in the log file I noticed:

```
RAN SCRIPT
STDERR Traceback (most recent call last):
  File "/var/folders/h5/trswlc4n1056qz7d1bz9fx9h0000gn/T/ansible_IYDPZr/ansible_module_cloudformation.py", line 401, in <module>
    main()
  File "/var/folders/h5/trswlc4n1056qz7d1bz9fx9h0000gn/T/ansible_IYDPZr/ansible_module_cloudformation.py", line 269, in main
    template_body = open(module.params['template'], 'r').read()
IOError: [Errno 2] No such file or directory: 'secgroup.json'
```

You **must** put in the absolute path to the file.  I solved this by using:

```
  - name: launch lab stack
    local_action:
      module: cloudformation
      stack_name: "bug-test"
      state: present
      region: us-east-1
      disable_rollback: true
      template: "{{ playbook_dir }}/secgroup.json"
```

Not ideal, but works.  I think a bug should be filed in async_wrapper so it can better catch errors from the modules it calls.

**Comment by jsmartin** [comment]
I've put to make ansible a little more helpful in diagnosing these kinds of async-specific issues.  PR here: https://github.com/ansible/ansible-modules-core/pull/4591

Going to file this notabug .

**Comment by shadycuz** [comment]
@jsmartin, I am seeing this again, should I open a new issue?

**Comment by tima** [comment]
@shadycuz This issue has been closed for over a year and this repo locked and migrated to ansible/ansible.
